File: 04-hash-tables/01-basic-hash-table/csharp/HashTable.cs
/**
 * 雜湊表（Hash Table）- C# 實作
 * 使用鏈結法（chaining）處理碰撞
 *
 * Hash Table implementation using chaining for collision resolution
 */

using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace DataStructures
{
    /// <summary>
    /// 雜湊表泛型類別
    /// Generic Hash Table class
    /// </summary>
    /// <typeparam name="TKey">鍵的型別（key type）</typeparam>
    /// <typeparam name="TValue">值的型別（value type）</typeparam>
    public class HashTable<TKey, TValue> : IEnumerable<KeyValuePair<TKey, TValue>>
    {
        // ========== 內部類別 Inner Classes ==========

        /// <summary>
        /// 鍵值對節點
        /// Key-value pair node for chaining
        /// </summary>
        private class Entry
        {
            public TKey Key { get; set; }
            public TValue Value { get; set; }
            public Entry Next { get; set; }

            public Entry(TKey key, TValue value)
            {
                Key = key;
                Value = value;
                Next = null;
            }
        }

        // ========== 常數 Constants ==========
        private const int DefaultCapacity = 16;
        private const double MaxLoadFactor = 0.75;

        // ========== 成員變數 Member Variables ==========
        private Entry[] _buckets;       // 桶陣列 - Array of buckets
        private int _capacity;          // 桶的數量 - Number of buckets
        private int _size;              // 元素數量 - Number of elements

        // ========== 建構子 Constructors ==========

        /// <summary>
        /// 使用預設容量建立雜湊表
        /// Create hash table with default capacity
        /// </summary>
        public HashTable() : this(DefaultCapacity) { }

        /// <summary>
        /// 使用指定容量建立雜湊表
        /// Create hash table with specified capacity
        /// </summary>
        /// <param name="capacity">桶的數量（number of buckets）</param>
        public HashTable(int capacity)
        {
            if (capacity <= 0)
            {
                throw new ArgumentException(
                    "容量必須為正整數 / Capacity must be positive", nameof(capacity));
            }
            _capacity = capacity;
            _size = 0;
            _buckets = new Entry[capacity];
        }

        // ========== 屬性 Properties ==========

        /// <summary>
        /// 元素數量 - Number of elements
        /// </summary>
        public int Count => _size;

        /// <summary>
        /// 桶的數量 - Number of buckets
        /// </summary>
        public int Capacity => _capacity;

        /// <summary>
        /// 是否為空 - Is empty
        /// </summary>
        public bool IsEmpty => _size == 0;

        /// <summary>
        /// 負載因子 - Load factor
        /// </summary>
        public double LoadFactor => (double)_size / _capacity;

        // ========== 索引子 Indexer ==========

        /// <summary>
        /// 使用 [] 運算子存取
        /// Access using [] operator
        /// </summary>
        public TValue this[TKey key]
        {
            get
            {
                if (TryGetValue(key, out TValue value))
                {
                    return value;
                }
                throw new KeyNotFoundException($"Key not found: {key}");
            }
            set => Insert(key, value);
        }

        // ========== 私有方法 Private Methods ==========

        /// <summary>
        /// 計算雜湊索引
        /// Compute hash index
        /// </summary>
        private int Hash(TKey key)
        {
            if (key == null) return 0;
            // 確保回傳正數 - Ensure positive value
            return Math.Abs(key.GetHashCode()) % _capacity;
        }

        /// <summary>
        /// 擴容並重新雜湊
        /// Resize and rehash
        /// </summary>
        private void Rehash()
        {
            var oldBuckets = _buckets;
            _capacity *= 2;
            _buckets = new Entry[_capacity];
            _size = 0;

            // 重新插入所有元素 - Reinsert all elements
            foreach (var head in oldBuckets)
            {
                var current = head;
                while (current != null)
                {
                    Insert(current.Key, current.Value);
                    current = current.Next;
                }
            }
        }

        // ========== 公開方法 Public Methods ==========

        /// <summary>
        /// 插入鍵值對（若 key 已存在則更新）
        /// Insert key-value pair (update if key exists)
        ///
        /// 時間複雜度 Time Complexity: 平均 O(1), 最差 O(n)
        /// </summary>
        public void Insert(TKey key, TValue value)
        {
            int index = Hash(key);
            var current = _buckets[index];

            // 檢查 key 是否已存在 - Check if key exists
            while (current != null)
            {
                if (Equals(current.Key, key))
                {
                    current.Value = value;  // 更新 - Update existing
                    return;
                }
                current = current.Next;
            }

            // 新增節點到鏈結串列頭部 - Add new node at head
            var newEntry = new Entry(key, value)
            {
                Next = _buckets[index]
            };
            _buckets[index] = newEntry;
            _size++;

            // 檢查是否需要擴容 - Check if rehashing needed
            if (LoadFactor > MaxLoadFactor)
            {
                Rehash();
            }
        }

        /// <summary>
        /// 搜尋給定 key 的 value
        /// Search for value associated with key
        /// </summary>
        /// <returns>找到則回傳 value，否則回傳 default</returns>
        public TValue Search(TKey key)
        {
            TryGetValue(key, out TValue value);
            return value;
        }

        /// <summary>
        /// 嘗試取得給定 key 的 value
        /// Try to get value associated with key
        /// </summary>
        public bool TryGetValue(TKey key, out TValue value)
        {
            int index = Hash(key);
            var current = _buckets[index];

            while (current != null)
            {
                if (Equals(current.Key, key))
                {
                    value = current.Value;
                    return true;
                }
                current = current.Next;
            }

            value = default;
            return false;
        }

        /// <summary>
        /// 刪除指定的鍵值對
        /// Delete key-value pair
        /// </summary>
        /// <returns>刪除成功回傳 true，key 不存在回傳 false</returns>
        public bool Delete(TKey key)
        {
            int index = Hash(key);
            var current = _buckets[index];
            Entry prev = null;

            while (current != null)
            {
                if (Equals(current.Key, key))
                {
                    if (prev == null)
                    {
                        _buckets[index] = current.Next;
                    }
                    else
                    {
                        prev.Next = current.Next;
                    }
                    _size--;
                    return true;
                }
                prev = current;
                current = current.Next;
            }

            return false;
        }

        /// <summary>
        /// 檢查 key 是否存在
        /// Check if key exists
        /// </summary>
        public bool Contains(TKey key)
        {
            return TryGetValue(key, out _);
        }

        /// <summary>
        /// 清空雜湊表
        /// Clear all elements
        /// </summary>
        public void Clear()
        {
            _buckets = new Entry[_capacity];
            _size = 0;
        }

        /// <summary>
        /// 取得所有鍵
        /// Get all keys
        /// </summary>
        public IEnumerable<TKey> Keys
        {
            get
            {
                foreach (var head in _buckets)
                {
                    var current = head;
                    while (current != null)
                    {
                        yield return current.Key;
                        current = current.Next;
                    }
                }
            }
        }

        /// <summary>
        /// 取得所有值
        /// Get all values
        /// </summary>
        public IEnumerable<TValue> Values
        {
            get
            {
                foreach (var head in _buckets)
                {
                    var current = head;
                    while (current != null)
                    {
                        yield return current.Value;
                        current = current.Next;
                    }
                }
            }
        }

        // ========== IEnumerable 實作 ==========

        public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator()
        {
            foreach (var head in _buckets)
            {
                var current = head;
                while (current != null)
                {
                    yield return new KeyValuePair<TKey, TValue>(current.Key, current.Value);
                    current = current.Next;
                }
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        /// <summary>
        /// 字串表示
        /// String representation
        /// </summary>
        public override string ToString()
        {
            var sb = new StringBuilder("HashTable{");
            bool first = true;

            foreach (var kvp in this)
            {
                if (!first) sb.Append(", ");
                sb.Append($"{kvp.Key}={kvp.Value}");
                first = false;
            }

            sb.Append("}");
            return sb.ToString();
        }
    }

    // ========== 範例程式 Example Program ==========

    public class HashTableDemo
    {
        public static void Main(string[] args)
        {
            Console.WriteLine("=== 雜湊表基本操作示範 Hash Table Basic Operations Demo ===\n");

            // 建立雜湊表 - Create hash table
            var ht = new HashTable<string, int>();

            // 插入操作 - Insert operations
            Console.WriteLine("插入鍵值對 Inserting key-value pairs:");
            ht.Insert("apple", 100);
            ht.Insert("banana", 200);
            ht.Insert("cherry", 300);
            ht["date"] = 400;  // 使用索引子
            Console.WriteLine($"  雜湊表: {ht}");
            Console.WriteLine($"  大小: {ht.Count}, 負載因子: {ht.LoadFactor:F3}\n");

            // 搜尋操作 - Search operations
            Console.WriteLine("搜尋操作 Search operations:");
            Console.WriteLine($"  Search('apple') = {ht.Search("apple")}");
            Console.WriteLine($"  ht['banana'] = {ht["banana"]}");
            Console.WriteLine($"  Contains('cherry') = {ht.Contains("cherry")}");
            Console.WriteLine($"  Contains('grape') = {ht.Contains("grape")}\n");

            // 更新操作 - Update operation
            Console.WriteLine("更新操作 Update operation:");
            ht["apple"] = 150;
            Console.WriteLine("  更新 apple 的值為 150");
            Console.WriteLine($"  ht['apple'] = {ht["apple"]}\n");

            // 刪除操作 - Delete operation
            Console.WriteLine("刪除操作 Delete operations:");
            ht.Delete("banana");
            Console.WriteLine($"  刪除 'banana' 後: {ht}");
            Console.WriteLine($"  大小: {ht.Count}\n");

            // 迭代操作 - Iteration
            Console.WriteLine("迭代操作 Iteration:");
            Console.WriteLine($"  所有鍵 Keys: {string.Join(", ", ht.Keys)}");
            Console.WriteLine($"  所有值 Values: {string.Join(", ", ht.Values)}");
        }
    }
}

File: 04-hash-tables/01-basic-hash-table/java/HashTable.java
/** Doc block start
 * 雜湊表（Hash Table）- Java 實作
 * 使用鏈結法（chaining）處理碰撞
 *(blank line)
 * Hash Table implementation using chaining for collision resolution
 *(blank line)
 * @param <K> 鍵的型別（key type）
 * @param <V> 值的型別（value type）
 */  // End of block comment
public class HashTable<K, V> {  // Execute this statement as part of the data structure implementation.

    // ========== 內部類別 Inner Classes ==========

    /** Doc block start
     * 鍵值對節點 / Key-value pair node for chaining
     */  // End of block comment
    private static class Entry<K, V> {  // Execute this statement as part of the data structure implementation.
        K key;  // Execute this statement as part of the data structure implementation.
        V value;  // Execute this statement as part of the data structure implementation.
        Entry<K, V> next;  // Execute this statement as part of the data structure implementation.

        Entry(K key, V value) {  // Execute this statement as part of the data structure implementation.
            this.key = key;  // Assign or update a variable that represents the current algorithm state.
            this.value = value;  // Assign or update a variable that represents the current algorithm state.
            this.next = null;  // Assign or update a variable that represents the current algorithm state.
        }  // Close the current block scope.
    }  // Close the current block scope.

    // ========== 常數 Constants ==========
    private static final int DEFAULT_CAPACITY = 16;  // Assign or update a variable that represents the current algorithm state.
    private static final double MAX_LOAD_FACTOR = 0.75;  // Assign or update a variable that represents the current algorithm state.

    // ========== 成員變數 Member Variables ==========
    private Entry<K, V>[] buckets;  // 桶陣列 - Array of buckets
    private int capacity;            // 桶的數量 - Number of buckets
    private int size;                // 元素數量 - Number of elements

    // ========== 建構子 Constructors ==========

    /** Doc block start
     * 使用預設容量建立雜湊表 / Create hash table with default capacity
     */  // End of block comment
    public HashTable() {  // Execute this statement as part of the data structure implementation.
        this(DEFAULT_CAPACITY);  // Execute this statement as part of the data structure implementation.
    }  // Close the current block scope.

    /** Doc block start
     * 使用指定容量建立雜湊表 / Create hash table with specified capacity
     *(blank line)
     * @param capacity 桶的數量（number of buckets）
     */  // End of block comment
    @SuppressWarnings("unchecked")  // Execute this statement as part of the data structure implementation.
    public HashTable(int capacity) {  // Execute this statement as part of the data structure implementation.
        if (capacity <= 0) {  // Evaluate the condition and branch into the appropriate code path.
            throw new IllegalArgumentException(  // Throw an exception to signal an invalid argument or operation.
                "容量必須為正整數 / Capacity must be positive");  // Execute this statement as part of the data structure implementation.
        }  // Close the current block scope.
        this.capacity = capacity;  // Assign or update a variable that represents the current algorithm state.
        this.size = 0;  // Assign or update a variable that represents the current algorithm state.
        this.buckets = (Entry<K, V>[]) new Entry[capacity];  // Access or update the bucket storage used to hold entries or chains.
    }  // Close the current block scope.

    // ========== 私有方法 Private Methods ==========

    /** Doc block start
     * 計算雜湊索引 / Compute hash index
     *(blank line)
     * @param key 要雜湊的鍵
     * @return 桶的索引
     */  // End of block comment
    private int hash(K key) {  // Compute a hash-based index so keys map into the table's storage.
        if (key == null) {  // Evaluate the condition and branch into the appropriate code path.
            return 0;  // Return the computed result to the caller.
        }  // Close the current block scope.
        // 確保雜湊值為正數 - Ensure positive hash value
        return Math.abs(key.hashCode()) % capacity;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 擴容並重新雜湊 / Resize and rehash
     */  // End of block comment
    @SuppressWarnings("unchecked")  // Execute this statement as part of the data structure implementation.
    private void rehash() {  // Rehash entries into a larger table to keep operations near O(1) on average.
        Entry<K, V>[] oldBuckets = buckets;  // Access or update the bucket storage used to hold entries or chains.
        capacity *= 2;  // Assign or update a variable that represents the current algorithm state.
        buckets = (Entry<K, V>[]) new Entry[capacity];  // Access or update the bucket storage used to hold entries or chains.
        size = 0;  // Assign or update a variable that represents the current algorithm state.

        // 重新插入所有元素 - Reinsert all elements
        for (Entry<K, V> head : oldBuckets) {  // Iterate over a range/collection to process each item in sequence.
            Entry<K, V> current = head;  // Assign or update a variable that represents the current algorithm state.
            while (current != null) {  // Repeat while the loop condition remains true.
                insert(current.key, current.value);  // Execute this statement as part of the data structure implementation.
                current = current.next;  // Assign or update a variable that represents the current algorithm state.
            }  // Close the current block scope.
        }  // Close the current block scope.
    }  // Close the current block scope.

    // ========== 公開方法 Public Methods ==========

    /** Doc block start
     * 插入鍵值對（若 key 已存在則更新）
     * Insert key-value pair (update if key exists)
     *(blank line)
     * 時間複雜度 Time Complexity: 平均 O(1), 最差 O(n)
     *(blank line)
     * @param key 鍵
     * @param value 值
     */  // End of block comment
    public void insert(K key, V value) {  // Execute this statement as part of the data structure implementation.
        int index = hash(key);  // Compute a hash-based index so keys map into the table's storage.
        Entry<K, V> head = buckets[index];  // Access or update the bucket storage used to hold entries or chains.

        // 檢查 key 是否已存在 - Check if key exists
        Entry<K, V> current = head;  // Assign or update a variable that represents the current algorithm state.
        while (current != null) {  // Repeat while the loop condition remains true.
            if (current.key == null ? key == null : current.key.equals(key)) {  // Evaluate the condition and branch into the appropriate code path.
                current.value = value;  // 更新 - Update existing
                return;  // Return the computed result to the caller.
            }  // Close the current block scope.
            current = current.next;  // Assign or update a variable that represents the current algorithm state.
        }  // Close the current block scope.

        // 新增節點到鏈結串列頭部 - Add new node at head
        Entry<K, V> newEntry = new Entry<>(key, value);  // Assign or update a variable that represents the current algorithm state.
        newEntry.next = head;  // Assign or update a variable that represents the current algorithm state.
        buckets[index] = newEntry;  // Access or update the bucket storage used to hold entries or chains.
        size++;  // Execute this statement as part of the data structure implementation.

        // 檢查是否需要擴容 - Check if rehashing needed
        if (getLoadFactor() > MAX_LOAD_FACTOR) {  // Evaluate the condition and branch into the appropriate code path.
            rehash();  // Rehash entries into a larger table to keep operations near O(1) on average.
        }  // Close the current block scope.
    }  // Close the current block scope.

    /** Doc block start
     * 搜尋給定 key 的 value
     * Search for value associated with key
     *(blank line)
     * @param key 要搜尋的鍵
     * @return 找到則回傳 value，否則回傳 null
     */  // End of block comment
    public V search(K key) {  // Execute this statement as part of the data structure implementation.
        int index = hash(key);  // Compute a hash-based index so keys map into the table's storage.
        Entry<K, V> current = buckets[index];  // Access or update the bucket storage used to hold entries or chains.

        while (current != null) {  // Repeat while the loop condition remains true.
            if (current.key == null ? key == null : current.key.equals(key)) {  // Evaluate the condition and branch into the appropriate code path.
                return current.value;  // Return the computed result to the caller.
            }  // Close the current block scope.
            current = current.next;  // Assign or update a variable that represents the current algorithm state.
        }  // Close the current block scope.
        return null;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 刪除指定的鍵值對 / Delete key-value pair
     *(blank line)
     * @param key 要刪除的鍵
     * @return 刪除成功回傳 true，key 不存在回傳 false
     */  // End of block comment
    public boolean delete(K key) {  // Execute this statement as part of the data structure implementation.
        int index = hash(key);  // Compute a hash-based index so keys map into the table's storage.
        Entry<K, V> current = buckets[index];  // Access or update the bucket storage used to hold entries or chains.
        Entry<K, V> prev = null;  // Assign or update a variable that represents the current algorithm state.

        while (current != null) {  // Repeat while the loop condition remains true.
            if (current.key == null ? key == null : current.key.equals(key)) {  // Evaluate the condition and branch into the appropriate code path.
                if (prev == null) {  // Evaluate the condition and branch into the appropriate code path.
                    buckets[index] = current.next;  // Access or update the bucket storage used to hold entries or chains.
                } else {  // Handle the alternative branch when the condition is false.
                    prev.next = current.next;  // Assign or update a variable that represents the current algorithm state.
                }  // Close the current block scope.
                size--;  // Execute this statement as part of the data structure implementation.
                return true;  // Return the computed result to the caller.
            }  // Close the current block scope.
            prev = current;  // Assign or update a variable that represents the current algorithm state.
            current = current.next;  // Assign or update a variable that represents the current algorithm state.
        }  // Close the current block scope.
        return false;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 檢查 key 是否存在
     * Check if key exists
     *(blank line)
     * @param key 要檢查的鍵
     * @return 存在回傳 true，否則回傳 false
     */  // End of block comment
    public boolean contains(K key) {  // Execute this statement as part of the data structure implementation.
        return search(key) != null || (key == null && containsNullKey());  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 檢查是否包含 null 鍵
     * Check if contains null key
     */  // End of block comment
    private boolean containsNullKey() {  // Execute this statement as part of the data structure implementation.
        Entry<K, V> current = buckets[0];  // Access or update the bucket storage used to hold entries or chains.
        while (current != null) {  // Repeat while the loop condition remains true.
            if (current.key == null) {  // Evaluate the condition and branch into the appropriate code path.
                return true;  // Return the computed result to the caller.
            }  // Close the current block scope.
            current = current.next;  // Assign or update a variable that represents the current algorithm state.
        }  // Close the current block scope.
        return false;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 回傳元素數量 / Return number of elements
     */  // End of block comment
    public int size() {  // Execute this statement as part of the data structure implementation.
        return size;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 回傳桶的數量 / Return number of buckets
     */  // End of block comment
    public int capacity() {  // Execute this statement as part of the data structure implementation.
        return capacity;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 檢查是否為空 / Check if empty
     */  // End of block comment
    public boolean isEmpty() {  // Execute this statement as part of the data structure implementation.
        return size == 0;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 計算負載因子 / Calculate load factor
     */  // End of block comment
    public double getLoadFactor() {  // Execute this statement as part of the data structure implementation.
        return (double) size / capacity;  // Return the computed result to the caller.
    }  // Close the current block scope.

    /** Doc block start
     * 清空雜湊表 / Clear all elements
     */  // End of block comment
    @SuppressWarnings("unchecked")  // Execute this statement as part of the data structure implementation.
    public void clear() {  // Execute this statement as part of the data structure implementation.
        buckets = (Entry<K, V>[]) new Entry[capacity];  // Access or update the bucket storage used to hold entries or chains.
        size = 0;  // Assign or update a variable that represents the current algorithm state.
    }  // Close the current block scope.

    /** Doc block start
     * 字串表示 / String representation
     */  // End of block comment
    @Override  // Execute this statement as part of the data structure implementation.
    public String toString() {  // Execute this statement as part of the data structure implementation.
        StringBuilder sb = new StringBuilder("HashTable{");  // Assign or update a variable that represents the current algorithm state.
        boolean first = true;  // Assign or update a variable that represents the current algorithm state.

        for (Entry<K, V> head : buckets) {  // Iterate over a range/collection to process each item in sequence.
            Entry<K, V> current = head;  // Assign or update a variable that represents the current algorithm state.
            while (current != null) {  // Repeat while the loop condition remains true.
                if (!first) {  // Evaluate the condition and branch into the appropriate code path.
                    sb.append(", ");  // Execute this statement as part of the data structure implementation.
                }  // Close the current block scope.
                sb.append(current.key).append("=").append(current.value);  // Assign or update a variable that represents the current algorithm state.
                first = false;  // Assign or update a variable that represents the current algorithm state.
                current = current.next;  // Assign or update a variable that represents the current algorithm state.
            }  // Close the current block scope.
        }  // Close the current block scope.

        sb.append("}");  // Execute this statement as part of the data structure implementation.
        return sb.toString();  // Return the computed result to the caller.
    }  // Close the current block scope.

    // ========== 範例執行 Example Main ==========

    public static void main(String[] args) {  // Execute this statement as part of the data structure implementation.
        System.out.println("=== 雜湊表基本操作示範 Hash Table Basic Operations Demo ===\n");  // Execute this statement as part of the data structure implementation.

        // 建立雜湊表 - Create hash table
        HashTable<String, Integer> ht = new HashTable<>();  // Assign or update a variable that represents the current algorithm state.

        // 插入操作 - Insert operations
        System.out.println("插入鍵值對 Inserting key-value pairs:");  // Execute this statement as part of the data structure implementation.
        ht.insert("apple", 100);  // Execute this statement as part of the data structure implementation.
        ht.insert("banana", 200);  // Execute this statement as part of the data structure implementation.
        ht.insert("cherry", 300);  // Execute this statement as part of the data structure implementation.
        System.out.println("  雜湊表: " + ht);  // Execute this statement as part of the data structure implementation.
        System.out.printf("  大小: %d, 負載因子: %.3f%n%n", ht.size(), ht.getLoadFactor());  // Execute this statement as part of the data structure implementation.

        // 搜尋操作 - Search operations
        System.out.println("搜尋操作 Search operations:");  // Execute this statement as part of the data structure implementation.
        System.out.println("  search('apple') = " + ht.search("apple"));  // Assign or update a variable that represents the current algorithm state.
        System.out.println("  search('grape') = " + ht.search("grape"));  // Assign or update a variable that represents the current algorithm state.
        System.out.println("  contains('cherry') = " + ht.contains("cherry") + "\n");  // Assign or update a variable that represents the current algorithm state.

        // 更新操作 - Update operation
        System.out.println("更新操作 Update operation:");  // Execute this statement as part of the data structure implementation.
        ht.insert("apple", 150);  // Execute this statement as part of the data structure implementation.
        System.out.println("  更新 apple 的值為 150");  // Execute this statement as part of the data structure implementation.
        System.out.println("  search('apple') = " + ht.search("apple") + "\n");  // Assign or update a variable that represents the current algorithm state.

        // 刪除操作 - Delete operation
        System.out.println("刪除操作 Delete operations:");  // Execute this statement as part of the data structure implementation.
        ht.delete("banana");  // Execute this statement as part of the data structure implementation.
        System.out.println("  刪除 'banana' 後: " + ht);  // Execute this statement as part of the data structure implementation.
        System.out.println("  大小: " + ht.size());  // Execute this statement as part of the data structure implementation.
    }  // Close the current block scope.
}  // Close the current block scope.
